### PR TITLE
6 add image tag

### LINKED ISSUE #6: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [ ] When I add a new image, there is a form field to add tags.
- [ ] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [ ] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by steven0739**
Closes #6 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+if ActiveRecord.gem_version >= Gem::Version.new('5.0')
+  class ActsAsTaggableOnMigration < ActiveRecord::Migration[4.2]; end
+else
+  class ActsAsTaggableOnMigration < ActiveRecord::Migration; end
+end
+ActsAsTaggableOnMigration.class_eval do
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
Comment (Santhanaraman) [inline db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Can we add a not null constraint to the tag name column?

File: app/models/image.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 class Image < ApplicationRecord
   validates :url, presence: true
   validates_url :url, allow_blank: false
+  acts_as_taggable
```
Comment (Santhanaraman) [inline app/models/image.rb]: Can we add unit test to verify if the images are associated with tags?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -70,4 +70,15 @@ def test_create__fail
...
+
+  def test_create__tag_added
+    image_params = { url: 'https://learn.appfolio.com/apm/assets/homev3/features-carousel-slide-01-960x570-MEDIUM-LARGE.png',
+                     desc: 'best company ever',
+                     tag_list: 'tag here'}
+
+    post images_path, params: { image: image_params }
+
+    follow_redirect!
+    assert_includes @response.body, 'tag here'
```
Comment (Santhanaraman) [inline test/controllers/images_controller_test.rb]: Can we update other test cases to include tags in the params?

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -70,4 +70,15 @@ def test_create__fail
...
+  end
```
Comment (Santhanaraman) [inline test/controllers/images_controller_test.rb]: Can use `assert_select` and check if the correct html node has this tag?

File: app/views/images/index.html.erb [hunk 5]
```diff
@@ -12,6 +13,7 @@
     <tr>
       <td><%= image.desc %></td>
       <td><%= image_tag(image.url, height: '256', width: '256')%></td>
+      <td><%= image.tag_list %></td>
```
Comment (Santhanaraman) [inline app/views/images/index.html.erb]: Can we update the controller test and check if the tags are listed with appropriate image?

File: db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (steven0739) [reply db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: This is a file auto generated, so can we not change it?

File: app/models/image.rb (hunk 2 above)
Comment (steven0739) [reply app/models/image.rb]: added in image_test.rb

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (steven0739) [reply test/controllers/images_controller_test.rb]: updated

File: test/controllers/images_controller_test.rb (hunk 4 above)
Comment (steven0739) [reply test/controllers/images_controller_test.rb]: added

File: app/views/images/index.html.erb (hunk 5 above)
Comment (steven0739) [reply app/views/images/index.html.erb]: added, and all the modifications have been pushed

File: db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 1 above)
Comment (Santhanaraman) [reply db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Yeah we can. Its just that we are adding a constraint that makes sense for our case. Just to make it explicit that we don't support blank tag name

File: db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 6]
```diff
@@ -7,7 +7,7 @@ class ActsAsTaggableOnMigration < ActiveRecord::Migration; end
 ActsAsTaggableOnMigration.class_eval do
   def self.up
     create_table :tags do |t|
-      t.string :name
+      t.string :name, null: false
```
Comment (Santhanaraman) [inline db/migrate/20180630073755_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: The schema.rb still has the old column logic, can we rollback and migrate again? Also can we update the seeds to include the tags?

**Pull request merged**